File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        map <char, int> mp;

        int Max = 0, result = 0;
        int l = 0, r = 0;

        while (r < s.size()) {
            mp[s[r]]++;
            Max = max(Max, mp[s[r]]);

            if (r - l + 1 - Max > k) {
                mp[s[l]]--;
                l++;
            }
            result = max(result,r - l + 1);
            r++;
        }
        return result;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-5.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left, right = 1, max(piles)

        k, ans = 0, 0
        
        while left <= right:
            hour = 0
            k = (left + right) // 2
            for i in range(len(piles)):
                take = math.ceil(piles[i] / k)
                hour += take
                if hour > h:
                    left = k + 1
                    break
                if i == len(piles) - 1:
                    right = k - 1
                    ans = k

        return ans